core/tests/test_utils.py

```diff
@@ -80,6 +81,68 @@ def empty_environ():
         os.environ['HTTP_HOST'], os.environ['SERVER_PORT'])
 
 
+def get_filepath_from_filename(filename, rootdir):
+    """Returns filepath using the filename. Different files are present
+    in different subdirectories in the rootdir. So, we walk through the
+    rootdir and match the all the filenames with the given filename.
+    When a match is found the function returns the complete path of the
+    filename by using os.path.join(root, filename).
+
+    For example signup-page.mainpage.html is present in
+    core/templates/dev/head/pages/signup-page and error-page.mainpage.html is
+    present in core/templates/dev/head/pages/error-pages. So we walk through
+    core/templates/dev/head/pages and a match for signup-page.directive.html
+    is found in signup-page subdirectory and a match for
+    error-page.directive.html is found in error-pages subdirectory.
+
+    Args:
+        filename: str. The name of the file.
+        rootdir: str. The directory to search the file in.
+
+    Returns:
+        str | None. The path of the file if file is found otherwise
+            None.
+    """
+    # This is required since error files are served according to error status
+    # code. The file served is error-page.mainpage.html but it is compiled
+    # and stored as error-page-{status_code}.mainpage.html.
+    # So, we need to swap the name here to obtain the correct filepath.
+    if filename.startswith('error-page'):
+        filename = 'error-page.mainpage.html'
+    for root, _, files in os.walk(rootdir):
+        for name in files:
+            if name == filename:
+                return os.path.join(root, filename)
+    return None
+
+
+def mock_get_template(unused_self, filename):
+    """Mock for get_template function of jinja2 Environment. This mock is
+    required for backend tests since we do not have webpack compilation
+    before backend tests. The folder to search templates is webpack_bundles
+    which is generated after webpack compilation. Since this folder will be
+    missing, get_template function will return a TemplateNotFound error. So,
+    we use a mock for get_template which returns the html file from the source
+    directory instead.
+
+    Args:
+        unused_self: jinja2.environment.Environment. The Environment instance.
+        filename: str. The name of the file for which template is
+            to be returned.
+
+    Returns:
+        jinja2.environment.Template. The template for the given file.
+    """
+    filepath = get_filepath_from_filename(
+        filename, os.path.join('core', 'templates', 'dev', 'head', 'pages'))
+    if not filepath:
+        filepath = get_filepath_from_filename(
```

Comment:
I understand this mock retrieving from /pages. But why is it retrieving from /tests as well? That part seems weird.

Reply:
This is required since we have the `console_errors.html` template in `core/templates/dev/head/tests` folder. I have added a comment as well.

Reply:
Ah OK. Don't do it this way. /console_errors path is only needed for a protractor test, and the backend tests don't need this, right? There should be just one canonical directory for retrieving templates, so ideally get rid of these three lines of code altogether.

Reply:
The test [here](https://github.com/oppia/oppia/blob/develop/core/controllers/base_test.py#L107) checks this url.

Reply:
Exclude this URL explicitly from the test you linked (and explain why).

Reply:
Done.